rfcs/0051-hard-fork-package-generation.md

```diff
@@ -0,0 +1,148 @@
+
+To better support the workflow of having pre-packaged on-disk ledgers available to load for initial genesis ledger states, we will need to update the runtime configuration to support specifying ledgers via their merkle root instead of their set of accounts. The runtime configuration should continue to support both workflows, as the workflow of specifying specific accounts is very useful for development and testing, it is just undesirable when shipping a hard fork release.
+
+Specifically, we can update the `
+-->
+
+### Capturing Genesis State for Planned Hard Forks
+
+Under the current hard fork plan, a soft fork update will be shipped to the prior chain which bakes in shutdown logic at specific slot heights. At some slot height, the prior network will stop accepting transactions, and then at a subsequent slot height after that, the network will shutdown entirely (will stop producing/accepting/broadcasting blocks). At this point in time, we take the strongest chain, and within that chain, take the final block before the slot height we stopped accepting transactions at. This block is where we will dump the genesis state from.
+
+Rather than waiting to dump the genesis block data at the network halt slot, we will proactively dump every candidate hard fork genesis state, and select the correct candidate hard fork genesis state manually once the network halt slot is reached. There will be a special CLI flag that enables dumping the candidate hard fork genesis block states to disk automatically, which will be enabled on some nodes connected to the network. To ensure that we capture this data and do not lose it, a cron job will be executed on the nodes dumping this state which will attempt to upload any dumped candidate hard fork genesis block states to the cloud.
+
+__IMPORTANT NOTE:__ dumping ledgers back to back needs to not break a node (this may require additional work).
+
+### Capturing Genesis State for Unplanned Hard Forks
+
+In the case of an unplanned (emergency) hard fork, we do not have the liberty of allowing the daemon to automatically dump the state that we will start the new chain from. Additionally, the state we want to take could be arbitrarily far back in history (obviously we won't take something that is too far back in history since we don't want to undo much history, but we can't make assumptions about how far back we need to look).
+
+In such a circumstance, we must dump this state from the archive db rather than from an actively running daemon. The replayer tool is capable of materializing staged ledgers, snarked ledgers, and epoch ledgers for arbitrary blocks by replaying transactions from the archive db. It is important that this tool should take at most a couple of hours to execute, as the time spent executing this tool is a further delay in generating a new hard fork package for the emergency release.
```

Comment:
maybe we should have official checkpoints, so we can get to final state quicker? Is it technically possible or replayer has to start on final version of the db?

Reply:
+1. The replayer produces a checkpoint file after each run that is used in the next run to continue processing newer data. So, one would have to run a cron job to generate these checkpoint files which makes it possible to produce the required state within the 6 hours.